File: image-preset-builder/java/Preset.java
package com.contentstack.utils.presets;

import org.apache.http.client.utils.URIBuilder;
import org.jetbrains.annotations.NotNull;
import org.json.simple.JSONArray;
import org.json.simple.JSONObject;

import java.net.MalformedURLException;
import java.net.URISyntaxException;
import java.util.List;
import java.util.concurrent.atomic.AtomicReference;

import static com.contentstack.utils.presets.Constant.*;

public class Preset {

    public String resolvePresetByPresetName(
            @NotNull JSONObject asset,
            @NotNull String extensionUid,
            @NotNull String presetName) throws InvalidUrlException {

        final JSONArray localExtensionUid = returnExtensionId(asset, extensionUid);
        if (presetName.isEmpty()) {
            throw new IllegalArgumentException("Please provide valid Preset Name");
        }
        List<JSONObject> presetOptionKEYS = extractMetadata(localExtensionUid, presetName, "name");
        String assetUrl = (String) asset.get(URL);
        try {
            assetUrl = findPresetOptions(presetOptionKEYS, assetUrl);
        }catch (MalformedURLException | URISyntaxException e) {
            throw new InvalidUrlException(e.getLocalizedMessage());
        }
        return assetUrl;
    }

    String findPresetOptions(List<JSONObject> presetOptionKEYS, String assetUrl)
            throws MalformedURLException, URISyntaxException {
        AtomicReference<JSONObject> optionsJson = new AtomicReference<>();
        if (!presetOptionKEYS.isEmpty()) {
            presetOptionKEYS.forEach(options -> {
                if (options.containsKey(Constant.OPTIONS)) {
                    optionsJson.set((JSONObject) options.get(Constant.OPTIONS));
                }
            });
        }
        return getImageURL(assetUrl, optionsJson.get());
    }

    public String resolvePresetByPresetUID(
            @NotNull JSONObject asset,
            @NotNull String extensionUid,
            @NotNull String presetUid) {

        final JSONArray localExtensionUid = returnExtensionId(asset, extensionUid);
        List<JSONObject> presetOptionKEYS = extractMetadata(localExtensionUid, presetUid, "uid");
        String assetUrl = (String) asset.get(URL);
        try {
            assetUrl = findPresetOptions(presetOptionKEYS, assetUrl);
        } catch (Exception e) {
            e.printStackTrace();
        }
        return assetUrl;
    }

    private String getImageURL(String assetUrl, JSONObject optionKEYS)
            throws URISyntaxException, MalformedURLException {
        URIBuilder uriBuilder = new URIBuilder(assetUrl);
        if (optionKEYS.containsKey(TRANSFORM)) {
            JSONObject transform = (JSONObject) optionKEYS.get(Constant.TRANSFORM);
            getTransformParam(uriBuilder, transform);
        }
        if (optionKEYS.containsKey(IMAGE_TYPE)) {
            String imageType = (String) optionKEYS.get(Constant.IMAGE_TYPE);
            uriBuilder.addParameter("format", imageType);
        }
        if (optionKEYS.containsKey(QUALITY)) {
            String quality = (String) optionKEYS.get(Constant.QUALITY);
            uriBuilder.addParameter(Constant.QUALITY, quality);
        }
        if (optionKEYS.containsKey(EFFECTS)) {
            JSONObject effects = (JSONObject) optionKEYS.get(Constant.EFFECTS);
            getEffectsParams(uriBuilder, effects);
        }
        return uriBuilder.build().toURL().toString();
    }

    private void getTransformParam(URIBuilder uriBuilder, JSONObject transform) {
        if (transform.containsKey(HEIGHT)) {
            Object height = transform.get(Constant.HEIGHT);
            uriBuilder.addParameter(Constant.HEIGHT, height.toString());
        }
        if (transform.containsKey(WIDTH)) {
            Object width = transform.get(Constant.WIDTH);
            uriBuilder.addParameter(Constant.WIDTH, width.toString());
        }
        if (transform.containsKey(FLIP_MODE)) {
            String flipMode = (String) transform.get(Constant.FLIP_MODE);
            if (flipMode.equalsIgnoreCase("both")) {
                uriBuilder.addParameter(ORIENT, "3");
            }
            if (flipMode.equalsIgnoreCase("horiz")) {
                uriBuilder.addParameter(ORIENT, "2");
            }
            if (flipMode.equalsIgnoreCase("verti")) {
                uriBuilder.addParameter(ORIENT, "4");
            }
        }
    }

    private void getEffectsParams(URIBuilder uriBuilder, JSONObject effects) {
        if (effects.containsKey(BRIGHTNESS)) {
            uriBuilder.addParameter(BRIGHTNESS, effects.get(Constant.BRIGHTNESS).toString());
        }
        if (effects.containsKey(CONTRAST)) {
            uriBuilder.addParameter(CONTRAST, effects.get(Constant.CONTRAST).toString());
        }
        if (effects.containsKey(SHARPEN)) {
            uriBuilder.addParameter("saturation", effects.get("saturation").toString());
        }
        if (effects.containsKey(BLUR)) {
            uriBuilder.addParameter(BLUR, effects.get(Constant.BLUR).toString());
        }
    }

    private JSONArray returnExtensionId(JSONObject asset, String extensionUid) {
        JSONObject localMetadata = validator(asset);
        if (extensionUid.isEmpty()) {
            throw new IllegalArgumentException("Please provide valid extension uid");
        }
        JSONObject localExtensions = (JSONObject) localMetadata.get("extensions");
        return (JSONArray) localExtensions.get(extensionUid);
    }

}

File: image-preset-builder/python/preset/preset.py
def resolve(asset, extension_uid, preset_uid=None, preset_name=None):
    extensions = valid(asset)
    if isinstance(extensions, dict):
        if extension_uid in extensions:
            ext_uid_list = extensions[extension_uid]
        else:
            raise Exception("extension_uid not found in the extensions")
        for preset in ext_uid_list:
            items = preset['presets']
            for item in items:
                options = valid_preset_cred(item, preset_name, preset_uid)
                if options is not None:
                    return _executeOptions(options, asset['url'])


def result(params):
    from urllib.parse import urlencode
    return urlencode(params)


def _executeOptions(options, url):
    params = {}
    if 'transform' in options:
        params = _transformQuery(params, options['transform'])
    if 'image-type' in options:
        params['image-type'] = options['image-type']
    if 'quality' in options:
        params['quality'] = options['quality']
    if 'effects' in options:
        _effectQuery(params, options['effects'])
    return url + '?' + result(params)


def _transformQuery(params, transform):
    if 'height' in transform:
        params['height'] = transform['height']
    if 'width' in transform:
        params['width'] = transform['width']
    if 'flip-mode' in transform:
        flip_mode = transform['flip-mode']
        if flip_mode == "both":
            params['orient'] = '3'
        elif flip_mode == "horiz":
            params['orient'] = '2'
        elif flip_mode == "verti":
            params['orient'] = '4'
    return params


def _effectQuery(params, effects):
    if 'brightness' in effects:
        params['brightness'] = effects['brightness']
    elif 'contrast' in effects:
        params['contrast'] = effects['contrast']
    elif 'sharpen' in effects:
        params["saturation"] = effects["saturation"]
    elif 'blur' in effects:
        params['blur'] = effects['blur']
    return params


def valid(asset):
    if isinstance(asset, dict) and '_metadata' in asset:
        return asset['_metadata']['extensions']
    else:
        raise Exception("Invalid Asset, _metadata and _extensions not found")


def valid_ext(ext_uid):
    if isinstance(ext_uid, str) and ext_uid is not '':
        return ext_uid
    else:
        raise Exception("Please provide a valid extensions uid")


def valid_preset_cred(presets, preset_name, preset_uid):
    if preset_name is not None and not '' and preset_name == presets['name']:
        if not presets['options']:
            raise Exception("options not fount in preset object")
        return presets['options']
    elif preset_uid is not None and not '' and preset_uid == presets['uid']:
        if not presets['options']:
            raise Exception("options not fount in preset object")
        return presets['options']
    else:
        return None
